internal/cmd/cluster/service.go

```diff
@@ -0,0 +1,113 @@
+package cluster
+
+import (
+	"fmt"
+
+	ma "github.com/multiformats/go-multiaddr"
+	"github.com/urfave/cli/v2"
+	"github.com/wetware/ww/pkg/service"
+)
+
+func discover() *cli.Command {
+	return &cli.Command{
+		Name:    "discover",
+		Aliases: []string{"disc"},
+		Usage:   "discover a service",
+		Flags: []cli.Flag{
+			&cli.StringFlag{
+				Name:     "name",
+				Aliases:  []string{"n"},
+				Usage:    "service name",
+				Required: true,
+			},
+		},
+		Before: setup(),
+		After:  teardown(),
+		Action: discAction(),
+	}
+}
+
+func provide() *cli.Command {
+	return &cli.Command{
+		Name:    "provide",
+		Aliases: []string{"prov"},
+		Usage:   "provide a service",
+		Flags: []cli.Flag{
+			&cli.StringFlag{
+				Name:     "name",
+				Aliases:  []string{"n"},
+				Usage:    "service name",
+				Required: true,
+			},
+			&cli.StringSliceFlag{
+				Name:    "multiaddr",
+				Aliases: []string{"maddr"},
+				Usage:   "multiaddress of the service provdier",
+			},
+		},
+		Before: setup(),
+		After:  teardown(),
+		Action: provAction(),
+	}
+}
+
+func discAction() cli.ActionFunc {
+	return func(c *cli.Context) error {
+		disc, release := node.Service(c.Context)
+		defer release()
+
+		locator, release := disc.Locator(c.Context, c.String("name"))
+		defer release()
+
+		locs, release := locator.FindProviders(c.Context)
+		defer release()
+
+		for loc, ok := locs.Next(); ok; loc, ok = locs.Next() {
+			fmt.Println(loc.String())
+		}
+
+		return locs.Err()
+	}
+}
+
+func provAction() cli.ActionFunc {
+	return func(c *cli.Context) error {
+		disc, release := node.Service(c.Context)
+		defer release()
+
+		serviceId := c.String("name")
+
+		provider, release := disc.Provider(c.Context, serviceId)
+		defer release()
+
+		maddrsStr := c.StringSlice("maddr")
```

Comment:
API design suggestion:  I really like the `channel.Value` API, and I think we could make a really elegant API here as well using the same approach.

The idea would be to first define a `service.Value` type.  Note that it is a function.

```go
type Value func(ps api.Provider_provide_Params) error
```

Then, you create functions to assign specific types to one of the union fields.  So in this case:

```go
func Multiaddrs(ms ...ma.Multiaddr) Value {
    func(ps api.Provider_provide_Params) error {
        // TODO:  assign to ps
    }
}
```

You can also include a convenience like `MultiaddrStrings`, etc.

Then, in the `provAction()` body, you'd do this:

```go
disc, release := node.Service(c.Context)
defer release()

serviceId := c.String("name")

provider, release := disc.Provider(c.Context, serviceId)
defer release()

fut, release := provider.Provide(c.Context, service.MultiaddrStrings(c.Strings("maddr")))
// ...
```

Reply:
Mmmm, we can iterate over this in another PR. But I don't think this is that simple. Since we also need to pass other things such as `metadata`, and the `peerId`. Moreover, if when querying services we receive a struct of type `SignedLocatioin`, then to make it intuitive too, we should send provide `SignedLocation` structs